FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/R4FhirSpecificationProvider.cs

```diff
@@ -23,5 +81,145 @@ public bool IsValidFhirResourceType(string resourceType)
+                    compartment = parser.Parse<CompartmentDefinition>(compartmentContext);
+                }
+                catch (Exception exception)
+                {
+                    _logger.LogError($"Failed to parse compartment definition from file {compartmentFile}.");
+                    throw new FhirSpecificationProviderException($"Failed to parse compartment definition from file {compartmentFile}.", exception);
+                }
+
+                var compartmentType = compartment.Code;
+                if (compartmentType != null && IsValidCompartmentType(compartmentType.ToString()))
+                {
+                    var resourceTypes = compartment.Resource.Where(x => x.Param.Any()).Select(x => x.Code?.ToString()).ToHashSet();
+                    compartmentResourceTypesLookup.Add(compartmentType?.ToString(), resourceTypes);
+                }
+            }
+
+            return compartmentResourceTypesLookup;
+        }
+
+        private Tuple<Dictionary<string, HashSet<string>>, Dictionary<string, string>> BuildSearchParametersLookup()
+        {
+            string metaData = _dataClient.GetMetaDataAsync().Result;
+            var parser = new FhirJsonParser();
+
+            CapabilityStatement capabilityStatement;
+            try
+            {
+                capabilityStatement = parser.Parse<CapabilityStatement>(metaData);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError($"Failed to parse capability statement from FHIR server metadata.");
+                throw new FhirSpecificationProviderException($"Failed to parse capability statement from FHIR server metadata.", exception);
+            }
+
+            var resources = capabilityStatement.Rest.First().Resource;
+            var searchParameters = new Dictionary<string, HashSet<string>>();
+            var searchParameterIds = new Dictionary<string, string>();
+            foreach (var resource in resources)
+            {
+                var type = resource.Type?.ToString();
+                if (type != null)
+                {
+                    searchParameters[type] = resource.SearchParam.Select(x => x.Name).ToHashSet();
+                    resource.SearchParam.ForEach(x => searchParameterIds[SearchParameterKey(type, x.Name)] = x.Definition);
+                }
+            }
+
+            return new Tuple<Dictionary<string, HashSet<string>>, Dictionary<string, string>>(searchParameters, searchParameterIds);
+        }
+
+        /// <summary>
+        /// build SearchParameterDefinitionLookup, search parameter url to searchParameter object
+        /// </summary>
+        /// <returns>search parameter url to searchParameter object dictionary</returns>
+        private Dictionary<string, SearchParameter> BuildSearchParameterDefinitionLookup()
+        {
+            var parser = new FhirJsonParser();
+
+            var bundleContext = File.ReadAllText(_searchParameterFile);
```

Comment:
We may need handle errors when reading file?

Reply:
updated.